ambry-clustermap/src/main/java/com.github.ambry.clustermap/ClusterMapManager.java

```diff
@@ -261,45 +283,149 @@ private boolean checkEnoughUnallocatedRawCapacity(int replicaCountPerDatacenter,
+        Disk diskToAdd = Collections.max(diskCandidates, diskCapacityComparator);
+        disksToAllocate.add(diskToAdd);
+        nodesToExclude.add(diskToAdd.getDataNode());
+      }
+    }
+    return disksToAllocate;
+  }
 
-    for (Datacenter datacenter : dataCenters) {
-      int rcpd = replicaCountPerDatacenter;
-      Set<DataNode> nodesToExclude = new HashSet<DataNode>();
-      for (int i = 0; i < rcpd; i++) {
-        DataNode nodeWithMostCapacity = getDataNodeWithMostUnallocatedRawCapacity(datacenter, nodesToExclude);
-        Disk diskWithMostCapacity = getDiskWithMostUnallocatedRawCapacity(nodeWithMostCapacity, replicaCapacityInBytes);
-        allocatedDisks.add(diskWithMostCapacity);
-        nodesToExclude.add(nodeWithMostCapacity);
+  /**
+   * Return a list of disks for a new partition in the specified {@link Datacenter}. Retry a non rack-aware allocation
+   * in certain cases described below if {@code bestEffort} is enabled.
+   *
+   * @param replicaCountPerDatacenter how many replicas to attempt to allocate in the datacenter
+   * @param replicaCapacityInBytes the minimum amount of free space on a disk for a replica
+   * @param datacenter the {@link Datacenter} to allocate replicas in
+   * @param disableRackAwareness if {@code true}, do not attempt a rack-aware allocation
+   * @param bestEffort If {@code bestEffort} is {@code true} and {@code disableRackAwareness} is {@code false},
+   *                   this method will attempt a non rack-aware allocation if it cannot find a rack-aware allocation.
+   * @return a list of {@code replicaCountPerDatacenter} or fewer disks that can be allocated for a new partition in
+   *         the specified datacenter
+   */
+  private List<Disk> allocateDisksForPartitionWithRetry(int replicaCountPerDatacenter, long replicaCapacityInBytes,
+      Datacenter datacenter, boolean disableRackAwareness, boolean bestEffort) {
+
+    List<Disk> disks = allocateDisksForPartition(replicaCountPerDatacenter, replicaCapacityInBytes,
+        datacenter, disableRackAwareness);
+    if (datacenter.isRackAware() && !disableRackAwareness && (disks.size() < replicaCountPerDatacenter)) {
+      System.err.println("Rack-aware allocation failed for a partition on datacenter:" + datacenter.getName());
+      if (bestEffort) {
+        System.err.println("Attempting to perform a non rack-aware allocation.");
+        disks = allocateDisksForPartition(replicaCountPerDatacenter, replicaCapacityInBytes, datacenter, false);
       }
     }
-    return allocatedDisks;
+    if (disks.size() < replicaCountPerDatacenter) {
+      System.err.println("Could only allocate " + disks.size() + "/" + replicaCountPerDatacenter
+          + " replicas in datacenter: " + datacenter.getName());
+    }
+    return disks;
   }
 
-  // Best effort (or less) allocation of partitions. I.e., size of returned list may be less than numPartitions.
+
+  /**
+   * Perform a best effort (or less) allocation of partitions.
```

Comment:
looks like this does the best effort only if `bestEffort` is true? That is the right way to do it. Just need to change the documentation.